DDF-3211 Add configuration for Intrigue to directly retrieve map tiles.

troymohl opened the pull request:
#### What does this PR do?
Adds a configuration option to Intrigue for map layers to be directly retrieved instead of proxied through the DDF server.

#### Who is reviewing it? 
(please choose AT LEAST two reviewers that need to approve the PR before it can get merged)
@albany551  
@andrewkfiedler  

#### Select relevant component teams: 
https://github.com/orgs/codice/teams/ui

#### Choose 2 committers to review/merge the PR. 
(please choose ONLY two committers from below, delete the rest)
@bdeining
@millerw8

#### How should this be tested? (List steps with links to updated documentation)
Full build and install
Configure a map layer such that the proxy is disabled for the layer and verify in Chrome developer tools or similar tool that tile requests are made directly to the configured tile server.

Example:
  [{
    "alpha": 1,
    "parameters": {},
    "type": "OSM",
    "url": "http://c.tile.openstreetmap.org",
    "proxyEnabled": "false"
  }]

#### Any background context you want to provide?

#### What are the relevant tickets?
[DDF-3211](https://codice.atlassian.net/browse/DDF-3211)

#### Screenshots (if appropriate)

#### Checklist:
- [X] Documentation Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

bdeining commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/config/ConfigurationApplication.java:
```diff
@@ -397,8 +399,19 @@ private void setProxiesForImagery(List<Map<String, Object>> newImageryProviders)
-                    .toString()));
+            String baseUrl = newImageryProvider.get(URL).toString();
+            boolean proxyEnabled = true;
+            Object proxyEnabledProp = newImageryProvider.get(PROXY_ENABLED);
```
Can we use `instanceof` here?

bdeining commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/config/ConfigurationApplication.java:
```diff
@@ -397,8 +399,19 @@ private void setProxiesForImagery(List<Map<String, Object>> newImageryProviders)
+            String baseUrl = newImageryProvider.get(URL).toString();
+            boolean proxyEnabled = true;
+            Object proxyEnabledProp = newImageryProvider.get(PROXY_ENABLED);
+            if (proxyEnabledProp != null) {
```
Can we use instanceof here?

troymohl replied:
Fixed.

glenhein commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/config/ConfigurationApplication.java:
```diff
@@ -397,8 +399,19 @@ private void setProxiesForImagery(List<Map<String, Object>> newImageryProviders)
+            String baseUrl = newImageryProvider.get(URL).toString();
+            boolean proxyEnabled = true;
+            Object proxyEnabledProp = newImageryProvider.get(PROXY_ENABLED);
+            if (proxyEnabledProp instanceof String) {
```
how about simplifying this to `proxyEnabled = proxyEnabledProp != null && Boolean.parseBoolean(proxyEnabledProp.toString());` ?

troymohl replied:
Updated

troymohl replied:
Actually reverted this back to how it was. This is slightly different in that it will set it to false if the property is not there and this needs to default to `true`.

troymohl commented:
build now

troymohl commented:
build now

bdeining commented:
build now

millerw8 commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/config/ConfigurationApplication.java:
```diff
@@ -397,8 +399,19 @@ private void setProxiesForImagery(List<Map<String, Object>> newImageryProviders)
+            } else {
+                map.put(URL, baseUrl);
+            }
             proxiedImageryProviders.add(map);
```
Should we change the variable names to be more representative so that someone out of context can follow?  proxiedImageryProviders are really just imageryProviders now right, may or may not be proxied.  

And is the baseUrl really a baseUrl?  Or is it more the originalUrl or just url or imageryProviderUrl or providerUrl?

troymohl replied:
It's really the original URL. I will update the variable names. Unfortunately they were already using imageryProviders for the incoming JSON, so I'll pick another.

glenhein commented:
Ran `mvn clean install` and all tests and itests passed.

Configured an imagery provider: [{ "type": "WMS",  "url": "https://nowcoast.noaa.gov/arcgis/services/nowcoast/analysis_meteohydro_sfc_qpe_time/MapServer/WmsServer", "proxyEnabled" : "false", "layers": ["1"], "parameters": {"FORMAT": "image/png", "alpha": 1 }]

Without the proxyEnabled option, the following network activity was seen:

![screen shot 2017-08-18 at 9 07 37 am](https://user-images.githubusercontent.com/17837152/29467385-33fddb54-83f5-11e7-9aa7-8eced8fbd514.png)

With proxyEnabled set to false, the following network activity was seen:

![screen shot 2017-08-18 at 9 09 31 am](https://user-images.githubusercontent.com/17837152/29467409-4aaac5c4-83f5-11e7-82b4-c16540024492.png)

glenhein commented:
Here are two screenshots of the map showing that tiling is correct with and without proxying:

With Proxy:
![with-proxy](https://user-images.githubusercontent.com/17837152/29472270-13d7a0cc-8408-11e7-9b12-9e9d00bd9105.png)

Without Proxy:
![without-proxy](https://user-images.githubusercontent.com/17837152/29472271-13d86a5c-8408-11e7-96d3-7b839f1f2389.png)

troymohl commented:
@andrewkfiedler I updated Map Layer Wizard to include option for enabling/disabling proxy of map tiles. 
@millerw8 Correct there are no metatype changes as the Catalog UI configuration takes the map layer configuration as JSON. I have updated the .adoc to include an example with the proxyEnabled property. The Map Layer Wizard also supports enabling/disabling this property, and that should be the normal way it's configured.

troymohl commented:
build now

bdeining commented:
build now

bdeining commented:
Manually verified the configuration through the wizard proxied and un-proxied
![screen shot 2017-08-22 at 10 47 11 am](https://user-images.githubusercontent.com/6551038/29577092-72fab310-8727-11e7-8e60-b5d93512d0af.png)
![screen shot 2017-08-22 at 10 48 40 am](https://user-images.githubusercontent.com/6551038/29577109-830c2aae-8727-11e7-96bc-ec2e27811b06.png)

djblue commented on catalog/admin/module/catalog-admin-module-maplayers/src/main/webapp/map-layers/reducer.js:
```diff
@@ -161,6 +161,12 @@ export const validate = (providers) => {
 
+    const proxyEnabled = layer.get('proxyEnabled')
+
+    if (typeof proxyEnabled !== 'boolean') {
```
Did we want to add some tests for this validation?

djblue commented on catalog/admin/module/catalog-admin-module-maplayers/src/main/webapp/map-layers/reducer.js:
```diff
@@ -161,6 +161,12 @@ export const validate = (providers) => {
+    const proxyEnabled = layer.get('proxyEnabled')
+
+    if (typeof proxyEnabled !== 'boolean') {
+      errors = errors.setIn([i, 'proxyEnabled'], 'Proxy enabled settings needs to be true or false')
```
Should `settings` be singular?

bdeining commented:
build now
